File: leetcode/643-maximum-average-subarray-i.py
from typing import List

class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        window_sum = sum(nums[:k])
        max_sum = window_sum

        for i in range(1, len(nums) - k + 1):
            window_sum += nums[i + k - 1] - nums[i - 1]
            max_sum = max(max_sum, window_sum)

        return max_sum / k  # Return the maximum average

# Test cases
s = Solution()

print(
    12.75000 == s.findMaxAverage([1, 12, -5, -6, 50, 3], 4),
    -1.0 == s.findMaxAverage([-1], 1),
)

File: leetcode/643/643-maximum-average-subarray-i.go
package main

import (
	"fmt"
	"reflect"
)

func main() {
	got := findMaxAverage([]int{1, 12, -5, -6, 50, 3}, 4)
	want := 12.75
	// got := findMaxAverage([]int{0, 4, 0, 3, 2}, 1)
	// want := float64(4)
	if !reflect.DeepEqual(got, want) {
		fmt.Printf("test failed, expected %v, but got %v\n", want, got)
		return
	}

	fmt.Println("test passed")
}

func findMaxAverage(nums []int, k int) float64 {
	var (
		start          = 0
		end            = k - 1
		result float64 = 0
		max    float64 = 0
	)

	for end < len(nums) {
		result = sumSubArray(nums[start:end+1]) / float64(k)
		if max == 0 {
			max = result
		}
		if result > max && result != 0 {
			max = result
		}

		start++
		end++
	}

	return max
}

func sumSubArray(arr []int) float64 {
	sum := 0
	for _, v := range arr {
		sum += v
	}

	fmt.Println(sum)

	return float64(sum)
}
